File: Problem-032/solution.py
def pandigital_products(n):
    products = set()

    for a in range(1, 1000):
        for b in range(a, 10000):

            product = a * b

            combined = str(a) + str(b) + str(product)

            if len(combined) != n:
                continue

            if is_pandigital(combined, n):
                products.add(product)

    return sum(products)


def is_pandigital(s, n):

    # No zero allowed
    if "0" in s:
        return False

    # All digits must be unique
    if len(set(s)) != n:
        return False

    # Must contain 1 through n
    for i in range(1, n + 1):
        if str(i) not in s:
            return False

    return True

File: Problem-032/solution.js
function pandigitalProducts(n) {
    let products = new Set();

    for (let a = 1; a < 10000; a++) {
        for (let b = a; b < 10000; b++) {

            let product = a * b;

            let combined =
                String(a) +
                String(b) +
                String(product);

            if (combined.length !== n) {
                continue;
            }

            if (isPandigital(combined, n)) {
                products.add(product);
            }
        }
    }

    let total = 0;

    for (let product of products) {
        total += product;
    }

    return total;
}


function isPandigital(str, n) {


    if (str.includes("0")) {
        return false;
    }


    let digits = new Set(str);

    if (digits.size !== n) {
        return false;
    }


    for (let i = 1; i <= n; i++) {
        if (!digits.has(String(i))) {
            return false;
        }
    }

    return true;
}
